### PR TITLE
[RTM] Bring back the custom entry point

### PR DESCRIPTION
**Pull request by aschempp**
This will add support for legacy entry points. It worked fine with my local tests and the two entry points of Isotope eCommerce. Be aware that the legacy entry point does not support request-response, but I guess we have to live with that.

### CODE REVIEW

**Comment by aschempp** [comment]
Fixed tests and Scrutinizer. Let me know your thoughts before I finish stuff like unit tests.

File: src/Security/ContaoAuthenticator.php [hunk 1]
```diff
@@ -28,6 +30,21 @@
...
+     * @var ContainerInterface
+     */
+    private $container;
+
+    /**
+     * Constructor.
+     *
+     * @param ContainerInterface $container The service container
+     */
+    public function __construct(ContainerInterface $container)
```
Comment (leofeyer) [inline src/Security/ContaoAuthenticator.php]: This is an API change and cannot be done at the current point in time. We will have to extend `ContainerAware` instead.

**Comment by leofeyer** [comment]
Phew, these are a lot more changes than expected. Why did you have to adjust the framework and the authenticator class?

Given the amount of work it takes to make custom entry points work, I'd almost vote for not supporting them at all. Would this be a problem e.g. for Isotope?

File: src/Security/ContaoAuthenticator.php (hunk 1 above)
Comment (aschempp) [inline src/Security/ContaoAuthenticator.php]: I often hear @Toflar that constructors are not API. Not sure where that info is from though…

Comment (leofeyer) [inline src/Security/ContaoAuthenticator.php]: That's wrong. It was possible to instantiate the class without passing a container object before and it is no longer now. Definitely an API change.

**Comment by aschempp** [comment]
> Why did you have to adjust the framework and the authenticator class?

The authenticator incorrectly booted the framework regardless of the container scope. That means our "lazy" booting was not working at all (because Security is a `kernel.request` event).

The framework changes are minimal. The constants should not be set because they are already initialized in every entry point script.

File: src/Security/ContaoAuthenticator.php (hunk 1 above)
Comment (aschempp) [inline src/Security/ContaoAuthenticator.php]: True. Though we could also say this is a bug (see my other comment) in the RC ;-)

Comment (leofeyer) [inline src/Security/ContaoAuthenticator.php]: We should simply extend `ContainerAware` and handle the case `null === $this->container`. Problem solved without API change :)

Comment (aschempp) [inline src/Security/ContaoAuthenticator.php]: Not really. If the container is not there, we can't determine the scope, so not load the framework. That would change the whole class behavior.

Comment (Toflar) [inline src/Security/ContaoAuthenticator.php]: I certainly didn't say that. This is a BC break for sure.

Comment (leofeyer) [inline src/Security/ContaoAuthenticator.php]: > If the container is not there, we can't determine the scope, so not load the framework. That would change the whole class behavior.

I don't get it. Right now, there is no container, either, so how does the class behavior change?

Comment (aschempp) [inline src/Security/ContaoAuthenticator.php]: Currently it does load the user no matter the container scope. With the change, it would not, if the container is not injected (because that would be wrong).

**Comment by leofeyer** [comment]
@aschempp You did not answer my question if it would be a problem for Isotope if we would not support custom entry points in Contao 4 anymore.

**Comment by aschempp** [comment]
It certainly would. The currently implementation has two custom entry points. And I'm pretty sure there are quite a few other extensions. Just search the ER for files containing `define('TL_MODE'`

**Comment by aschempp** [comment]
Btw, I don't really see this thing as so complicated? If you remove the `ContaoFramework` and `ContaoAuthenticator` changes, it barely consists of a frontend controller and an entry point script.

**Comment by leofeyer** [comment]
I have fixed the `ContaoFramework` class in dcc2cf56cc08067ef8dbc854344949d92dcade08. I will fix the authenticator in a separate commit too, so we can see the actual changes of this PR.

**Comment by leofeyer** [comment]
I have extracted the authentication related changes in 9ce022922cffd7dca19fd71779c6bb2ec34017e3. Can you rebase the PR now please?

**Comment by aschempp** [comment]
I have rebased the PR

File: src/Controller/InitializeController.php [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+use Contao\CoreBundle\ContaoCoreBundle;
+use Sensio\Bundle\FrameworkExtraBundle\Configuration\Route;
+use Symfony\Bundle\FrameworkBundle\Controller\Controller;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+
+/**
+ * Handles the Contao initialize.php route.
+ *
+ * @author Andreas Schempp <https://github.com/aschempp>
```
Comment (Toflar) [inline src/Controller/InitializeController.php]: Shouldn't this whole class be marked as `@deprecated`?

File: src/Controller/InitializeController.php [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+        $masterRequest = $this->get('request_stack')->getMasterRequest();
+        $realRequest   = Request::createFromGlobals();
+        $scope         = ContaoCoreBundle::SCOPE_FRONTEND;
+
+        if (defined('TL_MODE') && TL_MODE === 'BE') {
+            $scope = ContaoCoreBundle::SCOPE_BACKEND;
+        }
+
+        // Necessary to make the base path correct
```
Comment (leofeyer) [inline src/Controller/InitializeController.php]: Why is this necessary?

File: src/Controller/InitializeController.php [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+     *
+     * @return Response
+     *
+     * @Route("/_initialize", name="contao_initialize")
+     */
+    public function indexAction()
+    {
+        trigger_error('Custom entry points are deprecated and will no longer work in Contao 5.0.', E_USER_DEPRECATED);
+
+        $masterRequest = $this->get('request_stack')->getMasterRequest();
```
Comment (leofeyer) [inline src/Controller/InitializeController.php]: Why is the master request not the real request?

File: src/Controller/InitializeController.php (hunk 3 above)
Comment (aschempp) [inline src/Controller/InitializeController.php]: The method `Request::getBasePath()` would return the path to the entry point script as relative URL all assets.

File: src/Controller/InitializeController.php (hunk 4 above)
Comment (aschempp) [inline src/Controller/InitializeController.php]: Because the master request is the one being generated in our `initialize.php` with the custom route.

File: src/Controller/InitializeController.php (hunk 3 above)
Comment (leofeyer) [inline src/Controller/InitializeController.php]: Shouldn't we replace the whole server bag with the one of the master request?

Comment (aschempp) [inline src/Controller/InitializeController.php]: Maybe that would work too…

Comment (aschempp) [inline src/Controller/InitializeController.php]: well no…

Comment (aschempp) [inline src/Controller/InitializeController.php]: The server bag of the master request is "incorrect" as well. The master request will point to the actual entry point script (which I think is good).

**Comment by leofeyer** [comment]
Ok, so this is RTM then, isn't it? What about the `if (defined('TL_SCRIPT'))` check in the framework initializer class? No longer required?

**Comment by aschempp** [comment]
Good point, they sure are still required! Must have got lost during the rebase.

**Comment by aschempp** [comment]
Fixed the `defined` check

**Comment by aschempp** [comment]
Found another problem and fixed it. Now the entry point works fine with the banner extension from @BugBuster1701

**Comment by leofeyer** [comment]
Don't forget to change the status to `RTM` once it is ready to merge.

**Comment by aschempp** [comment]
I just added another improvement. I had a very hard time yesterday to find the problem in af02d90843a6d0d9b352a1a4a5b7c141aea918d7 because the exception rendered by Symfony was not shown on the screen. This is RTM to me now.

**Comment by leofeyer** [comment]
Two more questions:

1. Why are you sending a response with the status code `100`? This seems completely wrong, because a valid response has a status code greater or equal to 200 and less than 300.

2. If the response does not have a status code of `100`, you execute `$kernel->terminate()` and `exit;`. But if the status code is `100`, you execute neither of them. Is this correct?

**Comment by aschempp** [comment]
1. From [Wikipedia](http://en.wikipedia.org/wiki/List_of_HTTP_status_codes)

    > **100 Continue**
This means that the server has received the request headers, and that the client should proceed to send the request body.

    I found this to be the most appropriate status code to tell our entry point script to continue the operation.

 2. Yes that's correct. If there is a Symfony response, we must send it and stop the script. Otherwise we leave the initialize.php and continue with the entry point script.

**Comment by leofeyer** [comment]
It is not appropriate at all (and not even implemented in HTTP/1.0). The status 100 is meant for checking in advance if a server is willing to accept a rather big POST request, instead of directly sending the POST request. A status 100 response does not contain a body, either!

We definitely need to find another way of identifying the responses generated by the `InitializeController` class. And before we call `$response->send()`, we **must** set the response code to `200`.

**Comment by aschempp** [comment]
You must have misread that. If the response code **is 100, we're not sending the response**. Therefore we don't need to set anything to *200*, and the actually used code does not matter at all for the client/server response. It's only a mean to communicate with the entry point script, because a controller **must** return a response.

**Comment by leofeyer** [comment]
Ok, I might have gotten that wrong. Still we should not use a wrong status code. Why don't you add an X-Status-Header or similar instead?

**Comment by aschempp** [comment]
The status code is not wrong… there’s just no status code for server-internal communication :D

We should use a status code that will certainly not be used by any other Symfony response, and I think that will be the case with a code 100.

**Comment by leofeyer** [comment]
It's still wrong. @discordier What do you think?

**Comment by leofeyer** [comment]
Merged in ffb9df4a4b5797531e55f9f77976d5dcca1ca5f3.
